src/main/scala/sigmastate/eval/CostingRules.scala

```diff
@@ -197,47 +197,53 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
     * This class defines generic costing helpers, to unify and simplify costing rules of individual methods.
     */
   abstract class Coster[T](obj: RCosted[T], method: SMethod, args: Seq[RCosted[_]]) {
-    def costOfArgs = (obj +: args).map(_.cost)
-    def sizeOfArgs = args.foldLeft(obj.size.dataSize)({ case (s, e) => s + e.size.dataSize })
+    def costOfArgs: Seq[Rep[Int]] = (obj +: args).map(_.cost)
+    def sizeOfArgs: Rep[Long] = args.foldLeft(obj.size.dataSize)({ case (s, e) => s + e.size.dataSize })
 
-    def constantSizeProperyAccess[R](prop: Rep[T] => Rep[R]): RCosted[R] =
-      withConstantSize(prop(obj.value), opCost(costOfArgs, selectFieldCost))
+    def constantSizePropertyAccess[R](prop: Rep[T] => Rep[R]): RCosted[R] = {
+      val value = prop(obj.value)
+      withConstantSize(value, opCost(value, costOfArgs, selectFieldCost))
+    }
 
-    def knownSizeProperyAccess[R](prop: Rep[T] => Rep[R], size: RSize[R]): RCosted[R] =
-      RCCostedPrim(prop(obj.value), opCost(costOfArgs, selectFieldCost), size)
+    def knownSizePropertyAccess[R](prop: Rep[T] => Rep[R], size: RSize[R]): RCosted[R] = {
+      val value = prop(obj.value)
+      RCCostedPrim(value, opCost(value, costOfArgs, selectFieldCost), size)
+    }
 
-    def knownLengthCollProperyAccess[R](prop: Rep[T] => Rep[Coll[R]], len: Rep[Int]): Rep[CostedColl[R]] =
-      mkCostedColl(prop(obj.value), len, opCost(costOfArgs, selectFieldCost))
+    def knownLengthCollPropertyAccess[R](prop: Rep[T] => Rep[Coll[R]], len: Rep[Int]): Rep[CostedColl[R]] =
+      mkCostedColl(prop(obj.value), len, opCost(obj.value, costOfArgs, selectFieldCost))
```

Comment:
@greenhat here you also need to introduce `val value = prop(obj.value)` and use it in opCost.
This means you should also add asserts to RCCostedColl, RCCostedPair, RCCostedOption to catch such cases.
The guiding principle is the following. Every `value` which is wrapped with RCCostedXXX, IF it comes with OpCost node, then this OpCost node should be unique for that specific value. 
This uniqueness is achievied by taking Id of the `value`.

Reply:
Thanks! Done.